File: framework/python/betterportal/sse.py
"""Scoped event delivery. LocalEvents serves one event loop, with no replica broadcast."""
from __future__ import annotations

import asyncio
from contextlib import asynccontextmanager, AbstractAsyncContextManager
from dataclasses import dataclass
import inspect
import json
from typing import AsyncGenerator, AsyncIterator, Awaitable, Callable, Generic, Protocol, TypeVar, cast

import anyvali as av
from .jsoncodec import loads


@dataclass(frozen=True)
class EventScope:
    tenant_id: str
    app_id: str


@dataclass(frozen=True)
class EventAddress:
    view_id: str
    scope: EventScope


class SubscriptionOverflow(RuntimeError):
    pass


class EventTransport(Protocol):
    async def publish(self, address: EventAddress, data: bytes) -> None: ...
    def subscribe(self, address: EventAddress) -> AbstractAsyncContextManager[AsyncIterator[bytes]]: ...
    async def aclose(self) -> None: ...


class _Subscription(AsyncIterator[bytes]):
    def __init__(self, capacity: int):
        self.queue: asyncio.Queue[bytes | None] = asyncio.Queue(capacity)
        self.closed = False
        self.error: Exception | None = None

    def finish(self, error: Exception | None = None) -> None:
        if self.closed:
            return
        self.closed, self.error = True, error
        while not self.queue.empty():
            self.queue.get_nowait()
        self.queue.put_nowait(None)

    async def __anext__(self) -> bytes:
        if self.closed and self.queue.empty():
            if self.error:
                raise self.error
            raise StopAsyncIteration
        value = await self.queue.get()
        if self.error:
            raise self.error
        if value is None:
            raise StopAsyncIteration
        return value


class LocalEvents:
    """Bounded ephemeral fan-out. Applications own and close the transport."""
    def __init__(self, *, capacity: int = 256, max_payload_bytes: int = 1024 * 1024):
        if capacity < 1 or max_payload_bytes < 1:
            raise ValueError("Event limits must be positive")
        self.capacity, self.max_payload_bytes = capacity, max_payload_bytes
        self._subscribers: dict[EventAddress, set[_Subscription]] = {}
        self._closed = False
        self._loop: asyncio.AbstractEventLoop | None = None

    def _check(self) -> None:
        if self._closed:
            raise RuntimeError("Event transport is closed")
        loop = asyncio.get_running_loop()
        if self._loop is None:
            self._loop = loop
        if self._loop is not loop:
            raise RuntimeError("LocalEvents belongs to another event loop")

    def _remove(self, address: EventAddress, subscriber: _Subscription) -> None:
        group = self._subscribers.get(address)
        if group is not None:
            group.discard(subscriber)
            if not group:
                self._subscribers.pop(address, None)

    async def publish(self, address: EventAddress, data: bytes) -> None:
        self._check()
        if not isinstance(data, bytes) or len(data) > self.max_payload_bytes:
            raise ValueError("Invalid event payload or byte limit exceeded")
        for subscriber in tuple(self._subscribers.get(address, ())):
            try:
                subscriber.queue.put_nowait(data)
            except asyncio.QueueFull:
                subscriber.finish(SubscriptionOverflow("SSE subscriber exceeded its pending event limit"))
                self._remove(address, subscriber)

    @asynccontextmanager
    async def subscribe(self, address: EventAddress) -> AsyncIterator[AsyncIterator[bytes]]:
        self._check()
        subscriber = _Subscription(self.capacity)
        self._subscribers.setdefault(address, set()).add(subscriber)
        try:
            yield subscriber
        finally:
            self._remove(address, subscriber)
            subscriber.finish()

    async def aclose(self) -> None:
        if self._closed:
            return
        self._check()
        self._closed = True
        for group in self._subscribers.values():
            for subscriber in group:
                subscriber.finish()
        self._subscribers.clear()


Input = TypeVar("Input")
Event = TypeVar("Event")
Context = TypeVar("Context")


def encode_event(data: str, *, event: str | None = None, event_id: str | None = None,
                 retry: int | None = None, max_data_bytes: int = 1024 * 1024) -> bytes:
    """Encode one UTF-8 SSE message. Hosts flush each yielded message."""
    if max_data_bytes < 1 or len(data.encode("utf-8")) > max_data_bytes:
        raise ValueError("SSE data byte limit exceeded")
    for value in (event, event_id):
        if value is not None and (any(c in value for c in "\r\n\0") or len(value.encode("utf-8")) > 1024):
            raise ValueError("Invalid SSE event name or ID")
    if retry is not None and (type(retry) is not int or retry < 0):
        raise ValueError("SSE retry must be a nonnegative integer")
    fields = [] if event is None else ["event: " + event]
    # splitlines() also splits Unicode separators, which SSE treats as data.
    fields.extend("data: " + line for line in data.replace("\r\n", "\n").replace("\r", "\n").split("\n"))
    if event_id is not None:
        fields.append("id: " + event_id)
    if retry is not None:
        fields.append("retry: " + str(retry))
    return ("\n".join(fields) + "\n\n").encode("utf-8")


class SseRoute(Generic[Input, Event, Context]):
    def __init__(self, view_id: str, input_schema: av.BaseSchema[Input], event_schema: av.BaseSchema[Event],
                 mapper: Callable[[Input, Context], Event | Awaitable[Event]], *, transport: EventTransport, max_payload_bytes: int = 1024 * 1024):
        if not view_id:
            raise ValueError("View ID is required")
        self.view_id, self.input_schema, self.event_schema = view_id, input_schema, event_schema
        self.mapper, self.transport = mapper, transport
        if max_payload_bytes < 1:
            raise ValueError("Payload limit must be positive")
        self.max_payload_bytes = max_payload_bytes

    def _encode(self, value: object) -> bytes:
        data = json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":")).encode("utf-8")
        if len(data) > self.max_payload_bytes:
            raise ValueError("Event payload byte limit exceeded")
        return data

    async def publish(self, scope: EventScope, value: Input) -> Input:
        parsed = self.input_schema.parse(value)
        payload = self._encode(parsed)
        await self.transport.publish(EventAddress(self.view_id, scope), payload)
        return parsed

    @asynccontextmanager
    async def subscribe(self, scope: EventScope, context: Context) -> AsyncIterator[AsyncIterator[Event]]:
        async with self.transport.subscribe(EventAddress(self.view_id, scope)) as subscription:
            async def events() -> AsyncGenerator[Event, None]:
                async for data in subscription:
                    if len(data) > self.max_payload_bytes:
                        raise ValueError("Event payload byte limit exceeded")
                    value = self.input_schema.parse(loads(data.decode("utf-8")))
                    mapped = self.mapper(value, context)
                    if inspect.isawaitable(mapped):
                        mapped = await mapped
                    event = self.event_schema.parse(cast(Event, mapped))
                    self._encode(event)
                    yield event
            iterator = events()
            try:
                yield iterator
            finally:
                await iterator.aclose()

    async def wire(self, scope: EventScope, context: Context, *,
                   render: Callable[[Event], str | Awaitable[str]] | None = None,
                   event: str | None = None) -> AsyncGenerator[bytes, None]:
        encode_event("", event=event, max_data_bytes=self.max_payload_bytes)
        async with self.subscribe(scope, context) as subscription:
            async for value in subscription:
                try:
                    data = render(value) if render else value if isinstance(value, str) else self._encode(value).decode("utf-8")
                    if inspect.isawaitable(data):
                        data = await data
                    message = encode_event(cast(str, data), event=event, max_data_bytes=self.max_payload_bytes)
                except Exception:
                    message = encode_event('{"code":"render_failed","message":"Event rendering failed"}', event="error", max_data_bytes=self.max_payload_bytes)
                yield message

File: framework/dotnet/BetterPortal/Sse.cs
using AnyVali;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Channels;

namespace BetterPortal;

public readonly record struct EventScope(string TenantId, string AppId);
public readonly record struct EventAddress(string ViewId, EventScope Scope);
public sealed class SubscriptionOverflowException() : Exception("SSE subscriber exceeded its pending event limit");

public interface IEventSubscription : IAsyncDisposable
{
    IAsyncEnumerable<ReadOnlyMemory<byte>> Read(CancellationToken cancellation = default);
}
public interface IEventTransport : IAsyncDisposable
{
    ValueTask Publish(EventAddress address, ReadOnlyMemory<byte> data, CancellationToken cancellation = default);
    ValueTask<IEventSubscription> Subscribe(EventAddress address, CancellationToken cancellation = default);
}

/// <summary>Thread-safe bounded fan-out in this process. No replica broadcast or event history.</summary>
public sealed class LocalEvents(int capacity = 256, int maxPayloadBytes = 1024 * 1024) : IEventTransport
{
    private readonly int capacity = capacity > 0 ? capacity : throw new ArgumentException("Capacity must be positive");
    private readonly int maxPayloadBytes = maxPayloadBytes > 0 ? maxPayloadBytes : throw new ArgumentException("Payload limit must be positive");
    private readonly object gate = new();
    private readonly Dictionary<EventAddress, HashSet<Subscription>> subscribers = [];
    private bool closed;

    private sealed class Subscription(LocalEvents owner, EventAddress address, int capacity) : IEventSubscription
    {
        // Disposal/overflow drains the queue, so reads are synchronized by Channel too.
        internal readonly Channel<ReadOnlyMemory<byte>> Queue = Channel.CreateBounded<ReadOnlyMemory<byte>>(new BoundedChannelOptions(capacity)
        { FullMode = BoundedChannelFullMode.Wait, SingleReader = false, SingleWriter = false, AllowSynchronousContinuations = false });
        internal void Finish(Exception? error = null)
        {
            while (Queue.Reader.TryRead(out _)) { }
            Queue.Writer.TryComplete(error);
        }
        public async IAsyncEnumerable<ReadOnlyMemory<byte>> Read([EnumeratorCancellation] CancellationToken cancellation = default)
        {
            await foreach (var value in Queue.Reader.ReadAllAsync(cancellation)) yield return value;
        }
        public ValueTask DisposeAsync()
        {
            lock (owner.gate) { owner.Remove(address, this); Finish(); }
            return ValueTask.CompletedTask;
        }
    }

    private void Remove(EventAddress address, Subscription subscription)
    {
        if (!subscribers.TryGetValue(address, out var group)) return;
        group.Remove(subscription);
        if (group.Count == 0) subscribers.Remove(address);
    }
    public ValueTask Publish(EventAddress address, ReadOnlyMemory<byte> data, CancellationToken cancellation = default)
    {
        cancellation.ThrowIfCancellationRequested();
        if (data.Length > maxPayloadBytes) throw new ArgumentException("Event payload byte limit exceeded");
        var copy = data.ToArray();
        lock (gate)
        {
            ObjectDisposedException.ThrowIf(closed, this);
            foreach (var subscriber in subscribers.GetValueOrDefault(address)?.ToArray() ?? [])
            {
                if (subscriber.Queue.Writer.TryWrite(copy)) continue;
                subscriber.Finish(new SubscriptionOverflowException());
                Remove(address, subscriber);
            }
        }
        return ValueTask.CompletedTask;
    }
    public ValueTask<IEventSubscription> Subscribe(EventAddress address, CancellationToken cancellation = default)
    {
        cancellation.ThrowIfCancellationRequested();
        lock (gate)
        {
            ObjectDisposedException.ThrowIf(closed, this);
            var subscriber = new Subscription(this, address, capacity);
            if (!subscribers.TryGetValue(address, out var group)) subscribers[address] = group = [];
            group.Add(subscriber);
            return ValueTask.FromResult<IEventSubscription>(subscriber);
        }
    }
    public ValueTask DisposeAsync()
    {
        lock (gate)
        {
            if (closed) return ValueTask.CompletedTask;
            closed = true;
            foreach (var group in subscribers.Values) foreach (var subscriber in group) subscriber.Finish();
            subscribers.Clear();
        }
        return ValueTask.CompletedTask;
    }
}

public sealed class SseRoute<TInput, TEvent, TContext>(string viewId, Schema inputSchema, Schema eventSchema,
    Func<TInput, TContext, CancellationToken, ValueTask<TEvent>> mapper, IEventTransport transport, int maxPayloadBytes = 1024 * 1024)
{
    private readonly Func<TInput, TContext, CancellationToken, ValueTask<TEvent>> map = mapper;
    private readonly int maxPayloadBytes = maxPayloadBytes > 0 ? maxPayloadBytes : throw new ArgumentException("Payload limit must be positive");
    public string ViewId { get; } = !string.IsNullOrEmpty(viewId) ? viewId : throw new ArgumentException("View ID is required");
    public Schema InputSchema { get; } = inputSchema;
    public Schema EventSchema { get; } = eventSchema;

    public async ValueTask<TInput> Publish(EventScope scope, TInput value, CancellationToken cancellation = default)
    {
        var parsed = Contracts.Parse<TInput>(InputSchema, value);
        await transport.Publish(new(ViewId, scope), Encode(parsed), cancellation);
        return parsed;
    }
    private byte[] Encode(object? value)
    {
        var data = Encoding.UTF8.GetBytes(Json.Write(value));
        if (data.Length > maxPayloadBytes) throw new ArgumentException("Event payload byte limit exceeded");
        return data;
    }
    public async ValueTask<Subscription> Subscribe(EventScope scope, TContext context, CancellationToken cancellation = default) =>
        new(this, await transport.Subscribe(new(ViewId, scope), cancellation), context);

    public sealed class Subscription(SseRoute<TInput, TEvent, TContext> route, IEventSubscription subscription, TContext context) : IAsyncDisposable
    {
        public async IAsyncEnumerable<TEvent> Read([EnumeratorCancellation] CancellationToken cancellation = default)
        {
            await foreach (var data in subscription.Read(cancellation))
            {
                if (data.Length > route.maxPayloadBytes) throw new ArgumentException("Event payload byte limit exceeded");
                var input = Contracts.Parse<TInput>(route.InputSchema, Json.Read(new UTF8Encoding(false, true).GetString(data.Span)));
                var result = Contracts.Parse<TEvent>(route.EventSchema, await route.map(input, context, cancellation).AsTask().WaitAsync(cancellation));
                route.Encode(result);
                yield return result;
            }
        }
        public ValueTask DisposeAsync() => subscription.DisposeAsync();
    }
}
